src/pybind/mgr/pg_autoscaler/module.py

```diff
@@ -0,0 +1,466 @@
+        ps, root_map, pool_root = self._get_pool_status(osdmap, pools)
+
+        # Anyone in 'warn', set the health message for them and then
+        # drop them from consideration.
+        too_few = []
+        too_many = []
+        health_checks = {}
+
+        total_ratio = dict([(r, 0.0) for r in root_map.iterkeys()])
+        total_target_ratio = dict([(r, 0.0) for r in root_map.iterkeys()])
+        target_ratio_pools = dict([(r, []) for r in root_map.iterkeys()])
+
+        total_bytes = dict([(r, 0) for r in root_map.iterkeys()])
+        total_target_bytes = dict([(r, 0.0) for r in root_map.iterkeys()])
+        target_bytes_pools = dict([(r, []) for r in root_map.iterkeys()])
+
+        for p in ps:
+            total_ratio[p['crush_root_id']] += max(p['actual_capacity_ratio'],
+                                                   p['target_ratio'])
+            if p['target_ratio'] > 0:
+                total_target_ratio[p['crush_root_id']] += p['target_ratio']
+                target_ratio_pools[p['crush_root_id']].append(p['pool_name'])
+            total_bytes[p['crush_root_id']] += max(
+                p['actual_raw_used'],
+                p['target_bytes'] * p['raw_used_rate'])
+            if p['target_bytes'] > 0:
+                total_target_bytes[p['crush_root_id']] += p['target_bytes'] * p['raw_used_rate']
+                target_bytes_pools[p['crush_root_id']].append(p['pool_name'])
+            if p['subtree_capacity'] == 0:
+                self.log.debug('skipping empty subtree %s', cr_name)
+                continue
+            if not p['would_adjust']:
+                continue
+            if p['pg_autoscale_mode'] == 'warn':
+                msg = 'Pool %s has %d placement groups, should have %d' % (
+                    p['pool_name'],
+                    p['pg_num_target'],
+                    p['pg_num_final'])
+                if p['pg_num_final'] > p['pg_num_target']:
+                    too_few.append(msg)
+                else:
+                    too_many.append(msg)
+
+            if p['pg_autoscale_mode'] == 'on':
+                # Note that setting pg_num actually sets pg_num_target (see
+                # OSDMonitor.cc)
+                r = self.mon_command({
+                    'prefix': 'osd pool set',
+                    'pool': p['pool_name'],
+                    'var': 'pg_num',
+                    'val': str(p['pg_num_final'])
+                })
+
+                if r[0] != 0:
+                    # FIXME: this is a serious and unexpected thing,
+                    # we should expose it as a cluster log error once
+                    # the hook for doing that from ceph-mgr modules is
+                    # in.
+                    self.log.error("pg_num adjustment on {0} to {1} failed: {2}"
+                                   .format(p['pool_name'],
+                                           p['pg_num_final'], r))
+
+        if len(too_few):
```

Comment:
nit: for basic types like list and dict you can say if x: to check if it contains anything